godu.go

```diff
@@ -27,8 +30,19 @@ func main() {
 	if err != nil {
 		log.Fatalln(err.Error())
 	}
-	log.Printf("godu will walk through `%s` that might take up to few minutes\n", rootFolderName)
-	rootFolder := core.WalkFolder(rootFolderName, ioutil.ReadDir, getIgnoredFolders())
+	progress := new(int32)
+	finished := new(int32)
+	writer := uilive.New()
+	writer.Start()
+	go func() {
+		for atomic.LoadInt32(finished) != 1 {
```

Comment:
this is the only part that I think is a bit sub-optimal. I'm not using synchronizing primitives too often and I guess there might better way to tel the subroutine to finish

Reply:
Don't make it that complicated. Use a single `progress := make(chan int)`, send the number of files to it and simply close it once you're done scanning. Then sum up all the values in a for-range loop. See https://github.com/viktomas/godu/pull/39/files#diff-42d42a00a17ea3a6a746ad0773e55831R62 for a quite efficient solution.

Reply:
While the solution you suggested looks better in code, blocking all those file walking workers on the progress updater causes roughly 15% increase in time it takes to walk my home folder. Making the progress channel buffered helps only marginally.

Please let me know if you have any suggestions how to make the progress routine more performant. Otherwise I'll regress back to atomic int and just use boolean channel to signal the end of processing.

Reply:
Here's a slightly optimized version using channels:

```patch
diff --git a/core/file_walker.go b/core/file_walker.go
index 25f267b..05f9025 100644
--- a/core/file_walker.go
+++ b/core/file_walker.go
@@ -61,6 +61,7 @@ func WalkFolder(
 ) *File {
 	var wg sync.WaitGroup
 	c := make(chan bool, 2*runtime.NumCPU())
+	progress <- 1 // We're scanning the root folder, not really worth the effort to send this
 	root := walkSubFolderConcurrently(path, nil, ignoringReadDir(ignoredFolders, readDir), c, &wg, progress)
 	wg.Wait()
 	close(progress)
@@ -76,7 +77,6 @@ func walkSubFolderConcurrently(
 	wg *sync.WaitGroup,
 	progress chan<- int,
 ) *File {
-	progress <- 1
 	result := &File{}
 	entries, err := readDir(path)
 	if err != nil {
@@ -86,8 +86,15 @@ func walkSubFolderConcurrently(
 	dirName, name := filepath.Split(path)
 	result.Files = make([]*File, 0, len(entries))
 	var mutex sync.Mutex
+	var numSubfolders int
+	defer func() {
+		if numSubfolders > 0 {
+			progress <- numSubfolders
+		}
+	}()
 	for _, entry := range entries {
 		if entry.IsDir() {
+			numSubfolders++
 			subFolderPath := filepath.Join(path, entry.Name())
 			wg.Add(1)
 			go func() {
diff --git a/godu.go b/godu.go
index 2359658..6000b96 100644
--- a/godu.go
+++ b/godu.go
@@ -32,21 +32,26 @@ func main() {
 }
 
 func updateProgress(progress <-chan int) {
+	const interval = 50 * time.Millisecond
+
 	writer := uilive.New()
 	writer.Out = os.Stderr
 	writer.Start()
 	defer writer.Stop()
-	lastUpdate := time.Now()
-	totalFolders := 0
+
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	var total int
 	for {
-		folders, more := <-progress
-		if !more {
-			break
-		}
-		totalFolders += folders
-		if time.Since(lastUpdate) > 50*time.Millisecond {
-			fmt.Fprintf(writer, "Walked through %d folders\n", totalFolders)
-			lastUpdate = time.Now()
+		select {
+		case c, ok := <-progress:
+			if !ok {
+				return
+			}
+			total += c
+		case <-ticker.C:
+			fmt.Fprintf(writer, "Walked through %d folders\n", total)
 		}
 	}
 }

```

I've benchmarked it on OS X 10.11.6 against your atomic and unoptimized-channel versions using [`hyperfine`](https://github.com/sharkdp/hyperfine) with a folder containing 98927 sub-folders. Here are the results:

```sh
$ PATH=/some/path ; hyperfine "./godu-atomic $PATH" "./godu-channel $PATH" "./godu-channel-optimized $PATH"
Benchmark #1: ./godu-atomic /some/path
  Time (mean ± σ):     16.085 s ±  0.362 s    [User: 12.620 s, System: 51.553 s]
  Range (min … max):   15.592 s … 16.947 s    10 runs

Benchmark #2: ./godu-channel /some/path
  Time (mean ± σ):     16.597 s ±  0.170 s    [User: 14.145 s, System: 54.774 s]
  Range (min … max):   16.477 s … 17.031 s    10 runs

Benchmark #3: ./godu-channel-optimized /some/path
  Time (mean ± σ):     16.316 s ±  0.119 s    [User: 12.979 s, System: 53.018 s]
  Range (min … max):   16.149 s … 16.528 s    10 runs

Summary
  './godu-atomic /some/path' ran
    1.01 ± 0.02 times faster than './godu-channel-optimized /some/path'
    1.03 ± 0.03 times faster than './godu-channel /some/path'
```

So using atomics is only ~1% faster on my machine. I'd prefer the clean code over using complicated atomic constructs.

> Otherwise I'll regress back to atomic int and just use boolean channel to signal the end of processing.

If the channel's sole purpose is to aid in synchronizing program flows, then it's better to use a `chan struct{}` as it occupies no memory. See https://dave.cheney.net/2014/03/25/the-empty-struct.

Reply:
I find this solution very elegant. Thank you. I've put you as a co-author on the last comit since I pretty much just took your code. And I've learned about ticker :) the only thing I changed was instead of `defer`, I made the decision whether to send progress or not when I had the `numSubFolder` count.

Reply:
> I made the decision whether to send progress or not when I had the numSubFolder count.

Actually, the deferred function is run just before returning, so `numSubFolder` holds its final value when the function is executed. `defer`ing the function in this case has the benefit that all accesses to `numSubFolder` happen within the following 10 lines or so after declaration which improves code quality quite a bit.

Reply:
Good point, I've put the updating back where you had it, just created a function to encapsulate the condition to avoid spending too much space in the core routine of the app with updating which is peripheral functionality.